File: source/table_updater.cpp
#include "table_updater.hpp"
#include "version_manager.hpp"

namespace TableUpdater
{
    TableInfo getTableInfo(const std::string &version)
    {
        std::string url = "https://patch.esoul.kakaogames.com/Live/" + version + "/Table/const_data_version.json";
        std::string response = HttpClient::get(url);

        json data = json::parse(response);
        TableInfo info;
        info.version = data["version"];
        info.action = data["action"];
        return info;
    }

    bool checkAndUpdateLiveTables(const std::string &version)
    {

        // 获取服务器上的实际版本信息
        std::string versionUrl = std::format("https://patch.esoul.kakaogames.com/Live/{}/Table/const_data_version.json", version);
        std::println("检查版本URL: {}", versionUrl);

        TableInfo tableInfo = getTableInfo(version);
        if (tableInfo.version == 0)
        {
            std::println("\033[31m获取数据表信息失败\033[0m");
            return false;
        }
        std::println("服务器数据表版本: {}", tableInfo.version);

        fs::path table_info_path = "./table_info.json";
        json table_info;

        // 检查正式服数据表文件夹是否存在
        bool liveTableExist = fs::exists("../live_table") && !fs::is_empty("../live_table");
        // std::println("Live 数据表目录存在且非空: {}", liveTableExist ? "是" : "否");

        if (fs::exists(table_info_path))
        {
            std::ifstream file(table_info_path);
            table_info = json::parse(file);

            // if (table_info.contains("live"))
            // {
            //     std::println("当前保存的信息:");
            //     std::println("版本: {}", table_info["live"]["version"].get<std::string>());
            //     std::println("表版本: {}", table_info["live"]["tableVersion"].get<int>());
            // }

            // 比较版本号和哈希值
            if (table_info.contains("live") &&
                table_info["live"]["version"] == version &&
                table_info["live"]["tableVersion"] == tableInfo.version &&
                liveTableExist)
            {
                std::println("\033[32mLive 服务器数据表已是最新版本\033[0m");
                return false;
            }
            else
            {
                std::println("需要更新数据表:");
                if (table_info.contains("live"))
                {
                    if (table_info["live"]["tableVersion"] != tableInfo.version)
                    {
                        std::println("表版本不同: {} -> {}",
                                     table_info["live"]["tableVersion"].get<int>(), tableInfo.version);
                    }
                }
                else
                {
                    std::println("- table_info.json 中没有 live 信息");
                }
            }
        }
        else
        {
            std::println("table_info.json 不存在，将创建新文件");
        }

        // 构建下载链接
        std::string zipUrl = std::format("https://patch.esoul.kakaogames.com/Live/{}/Table/data_{}.zip",
                                         version, tableInfo.version);
        // std::println("下载URL: {}", zipUrl);
        std::string zipPath = "../data_" + std::to_string(tableInfo.version) + ".zip";

        // 下载数据表压缩包（智能多线程下载）
        if (!FileDownloader::downloadWithRetry(zipUrl, zipPath, 3, true))
        {
            return false;
        }

        if (!fs::exists("../live_table"))
        {
            fs::create_directory("../live_table");
        }
        std::string unzipCommand = "unzip -o " + zipPath + " -d ../live_table/ > /dev/null 2>&1";
        system(unzipCommand.c_str());

        fs::remove(zipPath);

        // 解密正式服数据表
        std::vector<uint8_t> key, iv;
        if (!CryptoUtils::deriveKeyAndIv(key, iv))
        {
            std::println("\033[31m密钥派生失败\033[0m");
            return false;
        }

        std::vector<fs::path> files_to_decrypt;
        for (const auto &entry : fs::directory_iterator("../live_table"))
        {
            if (entry.is_regular_file() && !CryptoUtils::isFileDecrypted(entry.path()))
            {
                files_to_decrypt.push_back(entry.path());
            }
        }

        if (!CryptoUtils::decryptFiles(files_to_decrypt, key, iv))
        {
            std::println("\033[31m正式服数据表解密失败\033[0m");
            return false;
        }

        // 转换正式服数据表
        if (!fs::exists("../live_json"))
        {
            fs::create_directory("../live_json");
        }
        if (!TableConverter::convertTablesToJson("../schema", "../live_table", "../live_json"))
        {
            std::println("\033[31m正式服数据表转换失败\033[0m");
            return false;
        }

        table_info["live"] = {
            {"version", version},
            {"tableVersion", tableInfo.version}};

        std::ofstream outFile(table_info_path);
        outFile << table_info.dump(4);

        std::println("\033[32mLive 服务器数据表已更新到最新版本\033[0m");
        return true;
    }

    bool checkReviewVersion(const std::string &version, std::string &cdnDate)
    {
        std::string url = "https://gc-infodesk-zinny3.kakaogames.com/v2/app?appId=743491&appVer=" +
                          version + "&market=googlePlay&sdkVer=1&os=android&lang=en";
        try
        {
            std::string response = HttpClient::get(url);
            if (response.empty())
            {
                return false;
            }

            json data;
            try
            {
                data = json::parse(response);
            }
            catch (const json::parse_error &e)
            {
                return false;
            }

            // 检查是否是review版本
            if (data.contains("content") &&
                data["content"].contains("appOption") &&
                data["content"]["appOption"].contains("appVerSvcStatus") &&
                data["content"]["appOption"]["appVerSvcStatus"] == "review")
            {
                // 确保cdnAddr存在且不为null
                if (data["content"]["appOption"].contains("cdnAddr") &&
                    !data["content"]["appOption"]["cdnAddr"].is_null())
                {
                    std::string cdnAddr = data["content"]["appOption"]["cdnAddr"];
                    std::regex date_regex(R"(/Review/(\d{4}))");
                    std::smatch matches;
                    if (std::regex_search(cdnAddr, matches, date_regex))
                    {
                        cdnDate = matches[1];
                        return true;
                    }
                }
            }
        }
        catch (const std::exception &e)
        {
            return false;
        }
        catch (...)
        {
            return false;
        }

        return false;
    }

    ReviewServerInfo checkReviewServer(const std::string &baseVersion)
    {
        ReviewServerInfo info;
        info.exists = false;
        info.version = "";
        info.cdnDate = "";
        info.tableInfo.version = 0;
        info.tableInfo.action = 0;

        // 首先搜索新版本
        std::vector<std::string> versions = VersionManager::generatePossibleVersions(baseVersion);

        // 获取CPU核心数，如果获取失败则使用默认值128
        const uint32_t cpu_cores = std::thread::hardware_concurrency();
        // 将线程数设置为核心数的18倍，但不超过1024
        const int max_threads = std::min(static_cast<uint32_t>(1024),
                                         cpu_cores > 0 ? cpu_cores * 18 : 128);

        std::vector<std::future<std::pair<bool, std::string>>> futures;
        std::vector<std::string> pending_versions;
        size_t total_versions = versions.size();
        size_t checked_versions = 0;
        static size_t last_output_length = 0;
        bool found_version = false;

        for (const auto &ver : versions)
        {
            if (found_version)
                break;
            if (futures.size() >= max_threads)
            {
                for (size_t i = 0; i < futures.size(); i++)
                {
                    auto result = futures[i].get();
                    checked_versions++;

                    if (result.first)
                    {
                        info.exists = true;
                        info.version = pending_versions[i];
                        info.cdnDate = result.second;
                        found_version = true;
                        break;
                    }
                    ProgressDisplay::update("检查进度", checked_versions, total_versions, ver, &last_output_length);
                }

                futures.clear();
                pending_versions.clear();
                if (found_version)
                    break;
            }

            futures.push_back(std::async(std::launch::async, [ver]()
                                         {
                std::string cdnDate;
                bool success = checkReviewVersion(ver, cdnDate);
                return std::make_pair(success, cdnDate); }));
            pending_versions.push_back(ver);
        }

        for (size_t i = 0; i < futures.size(); i++)
        {
            auto result = futures[i].get();
            checked_versions++;

            if (result.first)
            {
                info.exists = true;
                info.version = pending_versions[i];
                info.cdnDate = result.second;
                found_version = true;
                break;
            }


            ProgressDisplay::update("检查进度", checked_versions, total_versions, pending_versions[i], &last_output_length);
        }

        if (!found_version && checked_versions < total_versions)
        {
            checked_versions = total_versions;
            ProgressDisplay::update("检查进度", checked_versions, total_versions, versions.back(), &last_output_length);
        }
        ProgressDisplay::clear(last_output_length);

        if (!info.exists)
        {
            fs::path table_info_path = "./table_info.json";
            if (fs::exists(table_info_path))
            {
                try
                {
                    std::ifstream file(table_info_path);
                    json table_info = json::parse(file);

                    if (table_info.contains("review") &&
                        table_info["review"].contains("version") &&
                        !table_info["review"]["version"].is_null() &&
                        table_info["review"].contains("cdnDate") &&
                        !table_info["review"]["cdnDate"].is_null() &&
                        table_info["review"].contains("tableVersion") &&
                        !table_info["review"]["tableVersion"].is_null())
                    {

                        info.exists = true;
                        info.version = table_info["review"]["version"].get<std::string>();
                        info.cdnDate = table_info["review"]["cdnDate"].get<std::string>();
                        info.tableInfo.version = table_info["review"]["tableVersion"].get<int>();

                        // 验证这个版本是否可访问
                        std::string url = std::format(
                            "https://patch.esoul.kakaogames.com/Review/{}/{}/Table/const_data_version.json",
                            info.cdnDate, info.version);

                        std::string response = HttpClient::get(url);
                        if (!response.empty())
                        {
                            try
                            {
                                json data = json::parse(response);
                                info.tableInfo.action = data["action"].get<int>();
                            }
                            catch (...)
                            {
                            }
                        }
                    }
                }
                catch (const json::exception &e)
                {
                    std::println("\033[33mJSON解析错误: {}\033[0m", e.what());
                }
            }
        }

        if (!info.exists)
        {
            std::println("\033[33m未找到可用的 Review 服务器版本\033[0m");
        }

        return info;
    }

    /**
     * @brief 获取服务器数据表版本信息
     * @param reviewInfo Review服务器信息
     * @return 服务器数据表版本号，失败时返回-1
     */
    int getServerTableVersion(const ReviewServerInfo &reviewInfo)
    {
        std::string versionUrl = std::format("https://patch.esoul.kakaogames.com/Review/{}/{}/Table/const_data_version.json",
                                             reviewInfo.cdnDate, reviewInfo.version);
        std::println("检查版本URL: {}", versionUrl);

        std::string versionResponse = HttpClient::get(versionUrl);
        if (versionResponse.empty())
        {
            std::println("\033[31m获取 Review 服务器版本信息失败\033[0m");
            return -1;
        }

        try
        {
            json versionData = json::parse(versionResponse);
            int serverVersion = versionData["version"].get<int>();
            std::println("服务器数据表版本: {}", serverVersion);
            return serverVersion;
        }
        catch (const std::exception &e)
        {
            std::println("\033[31m解析版本信息失败: {}\033[0m", e.what());
            return -1;
        }
    }

    /**
     * @brief 检查是否需要更新Review数据表
     * @param reviewInfo Review服务器信息
     * @param serverTableVersion 服务器数据表版本
     * @return 需要更新返回true，否则返回false
     */
    bool needUpdateReviewTables(const ReviewServerInfo &reviewInfo, int serverTableVersion)
    {
        fs::path table_info_path = "./table_info.json";

        bool reviewTableExist = fs::exists("../review_table") && !fs::is_empty("../review_table");
        // std::println("Review 数据表目录存在且非空: {}", reviewTableExist ? "是" : "否");

        if (!fs::exists(table_info_path))
        {
            std::println("table_info.json 不存在，需要下载数据表");
            return true;
        }

        try
        {
            std::ifstream file(table_info_path);
            json table_info = json::parse(file);

            if (table_info.contains("review"))
            {
                // std::println("当前保存的信息:");
                // std::println("版本: {}", table_info["review"]["version"].get<std::string>());
                // std::println("表版本: {}", table_info["review"]["tableVersion"].get<int>());

                if (table_info["review"]["version"] == reviewInfo.version &&
                    table_info["review"]["cdnDate"] == reviewInfo.cdnDate &&
                    table_info["review"]["tableVersion"] == serverTableVersion &&
                    reviewTableExist)
                {
                    std::println("\033[32mReview 服务器数据表已是最新版本\033[0m");
                    return false;
                }

                std::println("需要更新数据表:");
                if (table_info["review"]["tableVersion"] != serverTableVersion)
                {
                    std::println("表版本不同: {} -> {}",
                                 table_info["review"]["tableVersion"].get<int>(), serverTableVersion);
                }
            }
            else
            {
                std::println("- table_info.json 中没有 review 信息");
            }
        }
        catch (const std::exception &e)
        {
            std::println("\033[31m解析table_info.json失败: {}\033[0m", e.what());
        }

        return true;
    }

    /**
     * @brief 下载并解压Review数据表
     * @param reviewInfo Review服务器信息
     * @param serverTableVersion 服务器数据表版本
     * @return 成功返回true，失败返回false
     */
    bool downloadReviewTables(const ReviewServerInfo &reviewInfo, int serverTableVersion)
    {
        // 构建下载链接
        std::string zipUrl = std::format("https://patch.esoul.kakaogames.com/Review/{}/{}/Table/data_{}.zip",
                                         reviewInfo.cdnDate, reviewInfo.version, serverTableVersion);
        // std::println("下载URL: {}", zipUrl);
        std::string zipPath = "../review_data_" + std::to_string(serverTableVersion) + ".zip";

        // 下载数据表压缩包（智能多线程下载）
        if (!FileDownloader::downloadWithRetry(zipUrl, zipPath, 3, true))
        {
            return false;
        }

        // 创建目录并解压
        if (!fs::exists("../review_table"))
        {
            fs::create_directory("../review_table");
        }

        std::string unzipCommand = "unzip -o " + zipPath + " -d ../review_table/ > /dev/null 2>&1";
        int result = system(unzipCommand.c_str());
        fs::remove(zipPath);

        if (result != 0)
        {
            std::println("\033[31m解压失败\033[0m");
            return false;
        }

        return true;
    }

    /**
     * @brief 解密Review数据表文件
     * @return 成功返回true，失败返回false
     */
    bool decryptReviewTables()
    {
        std::vector<uint8_t> key, iv;
        if (!CryptoUtils::deriveKeyAndIv(key, iv))
        {
            std::println("\033[31m密钥派生失败\033[0m");
            return false;
        }

        std::vector<fs::path> files_to_decrypt;
        for (const auto &entry : fs::directory_iterator("../review_table"))
        {
            if (entry.is_regular_file() && !CryptoUtils::isFileDecrypted(entry.path()))
            {
                files_to_decrypt.push_back(entry.path());
            }
        }

        if (!files_to_decrypt.empty())
        {
            if (!CryptoUtils::decryptFiles(files_to_decrypt, key, iv))
            {
                std::println("\033[31mReview 数据表解密失败\033[0m");
                return false;
            }
        }

        return true;
    }

    /**
     * @brief 转换Review数据表为JSON格式
     * @return 成功返回true，失败返回false
     */
    bool convertReviewTablesToJson()
    {
        if (!fs::exists("../review_json"))
        {
            fs::create_directory("../review_json");
        }

        if (!TableConverter::convertTablesToJson("../schema", "../review_table", "../review_json"))
        {
            std::println("\033[31mReview 数据表转换失败\033[0m");
            return false;
        }

        return true;
    }

    /**
     * @brief 更新table_info.json文件
     * @param reviewInfo Review服务器信息
     * @param serverTableVersion 服务器数据表版本
     * @return 成功返回true，失败返回false
     */
    bool updateTableInfoFile(const ReviewServerInfo &reviewInfo, int serverTableVersion)
    {
        try
        {
            fs::path table_info_path = "./table_info.json";
            json table_info;

            if (fs::exists(table_info_path))
            {
                std::ifstream file(table_info_path);
                table_info = json::parse(file);
            }

            table_info["review"] = {
                {"version", reviewInfo.version},
                {"cdnDate", reviewInfo.cdnDate},
                {"tableVersion", serverTableVersion}};

            std::ofstream outFile(table_info_path);
            outFile << table_info.dump(4);

            return true;
        }
        catch (const std::exception &e)
        {
            std::println("\033[31m更新table_info.json失败: {}\033[0m", e.what());
            return false;
        }
    }

    bool downloadAndProcessReviewTables(const ReviewServerInfo &reviewInfo)
    {
        int serverTableVersion = getServerTableVersion(reviewInfo);
        if (serverTableVersion == -1)
        {
            return false;
        }

        if (!needUpdateReviewTables(reviewInfo, serverTableVersion))
        {
            return false; // 已是最新版本，不需要更新
        }

        if (!downloadReviewTables(reviewInfo, serverTableVersion))
        {
            return false;
        }

        if (!decryptReviewTables())
        {
            return false;
        }

        if (!convertReviewTablesToJson())
        {
            return false;
        }

        if (!updateTableInfoFile(reviewInfo, serverTableVersion))
        {
            return false;
        }

        std::println("\033[32mReview 服务器数据表已更新到最新版本\033[0m");
        return true;
    }
}

File: Src/Python/table_updater.py
"""
数据表更新模块
"""

import json
import re
import subprocess
import shutil
import threading
from pathlib import Path
from enum import Enum
from dataclasses import dataclass
from typing import Optional
from concurrent.futures import ThreadPoolExecutor, as_completed
import multiprocessing

from rich.console import Console
from rich.progress import Progress, BarColumn, TextColumn, TimeRemainingColumn

from http_client import HttpClient
from file_downloader import FileDownloader
from crypto_utils import CryptoUtils
from table_converter import TableConverter
from version_manager import VersionManager

console = Console()


class ServerType(Enum):
    """服务器类型枚举"""

    GLOBAL_LIVE = "GlobalLive"
    GLOBAL_REVIEW = "GlobalReview"
    CN_LIVE = "CnLive"


@dataclass
class TableInfo:
    """数据表信息"""

    version: int = 0
    action: int = 0


@dataclass
class ReviewServerInfo:
    """Review服务器信息"""

    exists: bool = False
    version: str = ""
    cdn_date: int = 0
    table_info: TableInfo = None

    def __post_init__(self):
        if self.table_info is None:
            self.table_info = TableInfo()


class TableUpdater:
    """数据表更新类"""

    @staticmethod
    def get_table_info(version: str) -> TableInfo:
        """
        获取指定版本的数据表信息

        Args:
            version: 游戏版本号

        Returns:
            包含数据表版本和操作信息的结构体
        """
        url = f"https://patch.esoul.kakaogames.com/Live/{version}/Table/const_data_version.json"
        response = HttpClient.get(url)

        data = json.loads(response)
        info = TableInfo()
        info.version = data["version"]
        info.action = data["action"]
        return info

    @staticmethod
    def check_review_version(
        version: str, stop_event: Optional[threading.Event] = None
    ) -> tuple[bool, int]:
        """
        检查指定版本是否为可用的Review服务器版本

        Args:
            version: 要检查的版本号
            stop_event: 停止事件，当其他线程找到版本时会设置此事件

        Returns:
            (是否可用, CDN日期)的元组
        """
        # 如果已经找到版本，直接返回
        if stop_event and stop_event.is_set():
            return False, 0

        url = (
            f"https://gc-infodesk-zinny3.kakaogames.com/v2/app?"
            f"appId=743491&appVer={version}&market=googlePlay&sdkVer=1&os=android&lang=en"
        )

        try:
            # 使用更短的超时时间和重试次数
            response = HttpClient.get(url, retries=2, timeout=5)
            if not response:
                return False, 0

            # 再次检查停止事件
            if stop_event and stop_event.is_set():
                return False, 0

            data = json.loads(response)

            # 检查是否是review版本
            if (
                data.get("content", {}).get("appOption", {}).get("appVerSvcStatus")
                == "review"
            ):

                cdn_addr = data.get("content", {}).get("appOption", {}).get("cdnAddr")
                if cdn_addr and cdn_addr != "null":
                    match = re.search(r"/Review/(\d{4})", cdn_addr)
                    if match:
                        cdn_date = int(match.group(1))
                        return True, cdn_date
        except Exception:
            return False, 0

        return False, 0

    @staticmethod
    def check_review_server(base_version: str, server_region: str) -> ReviewServerInfo:
        """
        检查Review服务器并获取相关信息

        Args:
            base_version: 基础版本号，用于生成可能的版本号
            server_region: 服务器区域（如 "Global", "Cn" 等）

        Returns:
            包含Review服务器信息的结构体
        """
        info = ReviewServerInfo()

        # 首先搜索新版本
        versions = VersionManager.generate_possible_versions(base_version)

        # 获取CPU核心数
        cpu_cores = multiprocessing.cpu_count()
        # 将线程数设置为核心数的18倍，但不超过1024
        max_threads = min(1024, cpu_cores * 18 if cpu_cores > 0 else 128)

        total_versions = len(versions)
        checked_versions = 0
        found_version = False

        # 创建停止事件，用于通知所有线程停止
        stop_event = threading.Event()

        with Progress(
            TextColumn("[bold blue]{task.description}"),
            BarColumn(),
            TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
            TextColumn("({task.completed}/{task.total})"),
            TextColumn("{task.fields[version]}"),
            console=console,
            transient=True,
        ) as progress:
            task = progress.add_task("检查进度", total=total_versions, version="")

            with ThreadPoolExecutor(max_workers=max_threads) as executor:
                future_to_version = {
                    executor.submit(
                        TableUpdater.check_review_version, ver, stop_event
                    ): ver
                    for ver in versions
                }

                for future in as_completed(future_to_version):
                    ver = future_to_version[future]
                    checked_versions += 1

                    try:
                        success, cdn_date = future.result()
                        if success:
                            info.exists = True
                            info.version = ver
                            info.cdn_date = cdn_date
                            found_version = True
                            # 设置停止事件，通知所有线程停止
                            stop_event.set()
                            # 更新进度条到100%
                            progress.update(task, completed=total_versions, version=ver)
                            break
                    except Exception:
                        pass

                    progress.update(task, advance=1, version=ver)

                    if found_version:
                        break

        if not info.exists:
            # 尝试从本地配置读取
            table_info_path = Path("./table_info.json")
            if table_info_path.exists():
                try:
                    with open(table_info_path, "r", encoding="utf-8") as f:
                        table_info = json.load(f)

                    if (
                        server_region in table_info
                        and "Review" in table_info[server_region]
                        and table_info[server_region]["Review"].get("version")
                        and table_info[server_region]["Review"].get("cdnDate")
                        and table_info[server_region]["Review"].get("tableVersion")
                    ):

                        info.exists = True
                        info.version = table_info[server_region]["Review"]["version"]
                        info.cdn_date = table_info[server_region]["Review"]["cdnDate"]
                        info.table_info.version = table_info[server_region]["Review"][
                            "tableVersion"
                        ]

                        # 验证这个版本是否可访问
                        url = (
                            f"https://patch.esoul.kakaogames.com/Review/{info.cdn_date}/"
                            f"{info.version}/Table/const_data_version.json"
                        )

                        response = HttpClient.get(url)
                        if response:
                            try:
                                data = json.loads(response)
                                info.table_info.action = data.get("action", 0)
                            except:
                                pass
                except Exception as e:
                    console.print(f"[bold yellow]JSON解析错误: {e}[/bold yellow]")

        if not info.exists:
            console.print("[bold yellow]未找到可用的 Review 服务器版本[/bold yellow]")

        return info

    @staticmethod
    def update_data_tables(
        server_type: ServerType,
        version: str = "",
        review_info: Optional[ReviewServerInfo] = None,
    ) -> bool:
        """
        统一的数据表更新函数

        Args:
            server_type: 服务器类型
            version: 版本号
            review_info: Review服务器信息（可选）

        Returns:
            如果数据表需要更新并成功更新则返回True，否则返回False
        """
        # 1. 准备阶段：获取版本信息、下载链接、目标路径等
        zip_url = ""
        table_version = 0
        current_version = version
        server_region = ""
        table_type = ""  # "Live" or "Review"
        target_dir = Path()
        schema_dir = Path()
        cdn_date = 0  # Only for Review

        if server_type == ServerType.GLOBAL_LIVE:
            server_region = "Global"
            table_type = "Live"
            target_dir = Path("../../Table/Global/Live")
            schema_dir = Path("../../FlatBuffers/Schema/Global")

            # 获取服务器上的实际版本信息
            version_url = f"https://patch.esoul.kakaogames.com/Live/{current_version}/Table/const_data_version.json"
            console.print(f"检查版本URL: {version_url}")

            info = TableUpdater.get_table_info(current_version)
            if info.version == 0:
                console.print("[bold red]获取数据表信息失败[/bold red]")
                return False
            table_version = info.version
            console.print(f"服务器数据表版本: {table_version}")

            zip_url = f"https://patch.esoul.kakaogames.com/Live/{current_version}/Table/data_{table_version}.zip"

        elif server_type == ServerType.CN_LIVE:
            server_region = "Cn"
            table_type = "Live"
            target_dir = Path("../../Table/Cn/Live")
            schema_dir = Path("../../FlatBuffers/Schema/Global")

            # 获取国服配置
            cn_config = VersionManager.get_cn_server_config()
            if not cn_config.is_valid:
                console.print("[bold red]获取国服配置失败[/bold red]")
                return False
            current_version = cn_config.version

            # 尝试从每个URL获取数据表版本信息
            working_base_url = ""
            for base_url in cn_config.download_urls:
                try:
                    version_url = (
                        f"{base_url}/{current_version}/Table/const_data_version.json"
                    )
                    console.print(f"检查版本URL: {version_url}")

                    response = HttpClient.get(version_url)
                    if response:
                        data = json.loads(response)
                        table_version = data["version"]
                        working_base_url = base_url
                        console.print(f"服务器数据表版本: {table_version}")
                        break
                except:
                    continue

            if table_version == 0 or not working_base_url:
                console.print("[bold red]所有URL都无法获取数据表信息[/bold red]")
                return False

            zip_url = (
                f"{working_base_url}/{current_version}/Table/data_{table_version}.zip"
            )

        elif server_type == ServerType.GLOBAL_REVIEW:
            server_region = "Global"
            table_type = "Review"
            target_dir = Path("../../Table/Global/Review")
            schema_dir = Path("../../FlatBuffers/Schema/Global")

            # 如果没有提供 review_info，则自动检查
            if not review_info:
                if not version:
                    console.print(
                        "[bold red]GlobalReview 需要提供 base_version 或 review_info[/bold red]"
                    )
                    return False
                review_info = TableUpdater.check_review_server(version, server_region)
                if not review_info.exists:
                    console.print(
                        "[bold yellow]未找到可用的 Review 服务器版本[/bold yellow]"
                    )
                    return False

            current_version = review_info.version
            cdn_date = review_info.cdn_date

            table_version = TableUpdater.get_server_table_version(review_info)
            if table_version == -1:
                return False

            zip_url = (
                f"https://patch.esoul.kakaogames.com/Review/{cdn_date}/"
                f"{current_version}/Table/data_{table_version}.zip"
            )

        # 2. 检查本地信息，是否需要更新
        table_info_path = Path("./table_info.json")
        table_info = {}
        table_exist = target_dir.exists() and any(target_dir.iterdir())

        if table_info_path.exists():
            try:
                with open(table_info_path, "r", encoding="utf-8") as f:
                    table_info = json.load(f)

                if (
                    server_region in table_info
                    and table_type in table_info[server_region]
                    and table_info[server_region][table_type]["version"]
                    == current_version
                    and table_info[server_region][table_type]["tableVersion"]
                    == table_version
                    and table_exist
                ):

                    if (
                        server_type == ServerType.GLOBAL_REVIEW
                        and "cdnDate" in table_info[server_region][table_type]
                        and table_info[server_region][table_type]["cdnDate"] != cdn_date
                    ):
                        # Continue update if cdnDate mismatch for Review
                        pass
                    else:
                        console.print(
                            f"[bold green]{server_region} {table_type} 数据表已是最新版本[/bold green]"
                        )
                        return False
                else:
                    # Just log/print reason if needed
                    if (
                        server_region in table_info
                        and table_type in table_info[server_region]
                    ):
                        if (
                            table_info[server_region][table_type]["tableVersion"]
                            != table_version
                        ):
                            console.print(
                                f"表版本不同: {table_info[server_region][table_type]['tableVersion']} -> {table_version}"
                            )
            except:
                pass
        else:
            console.print("table_info.json 不存在，将创建新文件")

        # 3. 下载
        zip_path = f"../../data_{server_region}_{table_type}_{table_version}.zip"
        if not FileDownloader.download_with_retry(zip_url, zip_path, 3, True):
            return False

        # 4. 清理旧目录并解压
        if target_dir.exists():
            shutil.rmtree(target_dir)
        target_dir.mkdir(parents=True, exist_ok=True)

        unzip_command = ["unzip", "-o", zip_path, "-d", str(target_dir) + "/"]
        subprocess.run(unzip_command, capture_output=True, text=True)
        Path(zip_path).unlink()

        # 5. 解密
        key, iv = CryptoUtils.derive_key_and_iv()

        files_to_decrypt = []
        for file_path in target_dir.iterdir():
            if file_path.is_file() and not CryptoUtils.is_file_decrypted(file_path):
                files_to_decrypt.append(file_path)

        if not CryptoUtils.decrypt_files(files_to_decrypt, key, iv):
            console.print("[bold red]数据表解密失败[/bold red]")
            return False

        # 6. 转换
        if not TableConverter.convert_tables_to_json(
            str(schema_dir), str(target_dir), str(target_dir)
        ):
            console.print("[bold red]数据表转换失败[/bold red]")
            return False

        # 7. 清理 .tbl
        for file_path in target_dir.iterdir():
            if file_path.is_file() and file_path.suffix == ".tbl":
                file_path.unlink()

        # 8. 更新 JSON
        try:
            if server_region not in table_info:
                table_info[server_region] = {}

            table_info[server_region][table_type] = {
                "version": current_version,
                "tableVersion": table_version,
            }

            if server_type == ServerType.GLOBAL_REVIEW and cdn_date > 0:
                table_info[server_region][table_type]["cdnDate"] = cdn_date

            with open(table_info_path, "w", encoding="utf-8") as f:
                json.dump(table_info, f, indent=4, ensure_ascii=False)
        except:
            pass

        console.print(
            f"[bold green]{server_region} {table_type} 数据表已更新到最新版本[/bold green]"
        )
        return True

    @staticmethod
    def get_server_table_version(review_info: ReviewServerInfo) -> int:
        """
        获取服务器数据表版本信息

        Args:
            review_info: Review服务器信息

        Returns:
            服务器数据表版本号，失败时返回-1
        """
        version_url = (
            f"https://patch.esoul.kakaogames.com/Review/{review_info.cdn_date}/"
            f"{review_info.version}/Table/const_data_version.json"
        )
        console.print(f"检查版本URL: {version_url}")

        version_response = HttpClient.get(version_url)
        if not version_response:
            console.print("[bold red]获取 Review 服务器版本信息失败[/bold red]")
            return -1

        try:
            version_data = json.loads(version_response)
            server_version = version_data["version"]
            console.print(f"服务器数据表版本: {server_version}")
            return server_version
        except Exception as e:
            console.print(f"[bold red]解析版本信息失败: {e}[/bold red]")
            return -1
